src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -683,82 +930,218 @@ void ShenandoahFreeSet::clear() {
-               byte_size_in_proper_unit(region->used()), proper_unit_for_byte_size(region->used()));
+          byte_size_in_proper_unit(region->used()), proper_unit_for_byte_size(region->used()));
+      }
     }
   }
+}
+
+void ShenandoahFreeSet::rebuild() {
+  shenandoah_assert_heaplocked();
+  // This resets all state information, removing all regions from all sets.
+  clear();
+
+  log_debug(gc, free)("Rebuilding FreeSet");
 
-  // Evac reserve: reserve trailing space for evacuations
+  // This places regions that have alloc_capacity into the old_collector set if they identify as is_old() or the
+  // mutator set otherwise.
+  find_regions_with_alloc_capacity();
+
+  // Evac reserve: reserve trailing space for evacuations, with regions reserved for old evacuations placed to the right
+  // of regions reserved of young evacuations.
+  size_t young_reserve, old_reserve;
   if (!_heap->mode()->is_generational()) {
-    size_t to_reserve = (_heap->max_capacity() / 100) * ShenandoahEvacReserve;
-    reserve_regions(to_reserve);
+    young_reserve = (_heap->max_capacity() / 100) * ShenandoahEvacReserve;
+    old_reserve = 0;
   } else {
-    size_t young_reserve = (_heap->young_generation()->max_capacity() / 100) * ShenandoahEvacReserve;
-    // Note that all allocations performed from old-gen are performed by GC, generally using PLABs for both
+    // All allocations taken from the old collector set are performed by GC, generally using PLABs for both
     // promotions and evacuations.  The partition between which old memory is reserved for evacuation and
-    // which is reserved for promotion is enforced using thread-local variables that prescribe intentons within
-    // each PLAB.  We do not reserve any of old-gen memory in order to facilitate the loaning of old-gen memory
-    // to young-gen purposes.
-    size_t old_reserve = 0;
-    size_t to_reserve = young_reserve + old_reserve;
-    reserve_regions(to_reserve);
+    // which is reserved for promotion is enforced using thread-local variables that prescribe intentons for
+    // each PLAB's available memory.
+    if (_heap->has_evacuation_reserve_quantities()) {
+      // We are rebuilding at the end of final mark, having already established evacuation budgets for this GC pass.
+      young_reserve = _heap->get_young_evac_reserve();
+      old_reserve = _heap->get_promoted_reserve() + _heap->get_old_evac_reserve();
+    } else {
+      // We are rebuilding at end of GC, so we set aside budgets specified on command line (or defaults)
+      young_reserve = (_heap->young_generation()->max_capacity() * ShenandoahEvacReserve) / 100;
+      old_reserve = MAX2((_heap->old_generation()->max_capacity() * ShenandoahOldEvacReserve) / 100,
+                         ShenandoahOldCompactionReserve * ShenandoahHeapRegion::region_size_bytes());
+    }
   }
-
+  reserve_regions(young_reserve, old_reserve);
   recompute_bounds();
```

Comment:
If the `reserve_regions()` case is the only one where `recompute_bounds()` is called to batch the adjustments following mass movement and reshuffling of free regions into the three types of free sets, then what might work better (cf my previous comment about doing interval end-point adjustments incrementally whenever a set loses or receives a member rather than first add/remove one member and then do an `adjust_bounds_if_touched()`) would be to here have special versions of `add_` and `remove_` which affect the membership change but leave the intervals unchanged (as you do now), but only in the case of these mass adjustments of membership during a rebuild, then do a single sweep to set up the interval extrema at the end of the rebuild.

Reply:
What you suggest might be slightly more efficient, but the code is probably cleaner if we just remove "recompute_bounds()" and compute the bounds incrementally as part of the rebuild() and reserve() process.  This happens only twice per GC cycle (at start of evac, and after update-refs).  In the next revision of this PR, I will offer the "simplified" API.

Reply:
Actually, I did not remove recompute_bounds() because we currently depend on a side effect of this function, which is to decide whether to allocate old_colllector regions left-to-right or right-to-left.  I've placed a comment in the code to clarify a future TODO action.  Is it ok to leave as is for now?  (Or should I try to sort this out right now?)

Reply:
In the very latest review, I've removed recompute bounds entirely.  The only thing we do at the end of rebuild is decide whether to use right-to-left or left-to-right bias for allocation from OldCollector set.